src/components/analytics/TraceAnalytics.tsx

```diff
@@ -0,0 +1,35 @@
+import { sendAnalyticsEvent, Trace, TraceEvent } from '@uniswap/analytics'
+import { atomWithStorage, useAtomValue } from 'jotai/utils'
+import { memo } from 'react'
+
+const analyticsOptOutAtomName = 'optOutOfAnalytics'
+export const analyticsOptOutToggle = atomWithStorage<boolean>(analyticsOptOutAtomName, true)
+
+export const TraceAnalytics = memo((props: React.ComponentProps<typeof Trace>) => {
+  const allowAnalytics = useAtomValue(analyticsOptOutToggle)
+
+  return <Trace {...props} shouldLogImpression={allowAnalytics} />
+})
+
+TraceAnalytics.displayName = 'TraceAnalytics'
+
+export const TraceAnalyticsEvent = memo((props: React.ComponentProps<typeof TraceEvent>) => {
+  const allowAnalytics = useAtomValue(analyticsOptOutToggle)
+
+  return <TraceEvent {...props} shouldLogImpression={allowAnalytics} />
+})
+
+TraceAnalyticsEvent.displayName = 'TraceAnalyticsEvent'
+
+export const sendOptOutAnalyticsEvent: typeof sendAnalyticsEvent = (event, properties) => {
+  const value = localStorage.getItem(analyticsOptOutAtomName)
```

Comment:
Instead of referencing localStorage every time, could this be stored in-memory? Does it matter - ie, is accessing localStorage slow?

Reply:
its like a 10 thousandth of a millisecond to access local storage, so def not slow. The way to access this in-memory would be to use the useAtomValue hook in the component that calls this function then passing the value down, but that kinda defeats the purpose of this function